// rc-leaflet/lib/components/GeoJSON/PropTypes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prop_types_1 = __importDefault(require("prop-types"));
var PointType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['Point']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.number).isRequired
});
var MultiPointType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['MultiPoint']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.number)).isRequired
});
var LineStringType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['LineString']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.number)).isRequired
});
var MultiLineStringType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['MultiLineString']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.number))).isRequired
});
var PolygonType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['Polygon']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.number))).isRequired
});
var MultiPolygonType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['MultiPolygon']).isRequired,
    coordinates: prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.arrayOf(prop_types_1.default.number)))).isRequired
});
var GeometryType = prop_types_1.default.oneOfType([PointType, MultiPointType, LineStringType, MultiLineStringType, PolygonType, MultiPolygonType]);
exports.GeometryType = GeometryType;
var GeometryCollectionType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['GeometryCollection']).isRequired,
    coordinates: prop_types_1.default.arrayOf(GeometryType).isRequired
});
exports.GeometryCollectionType = GeometryCollectionType;
var FeatureType = prop_types_1.default.shape({
    id: prop_types_1.default.oneOfType([prop_types_1.default.string, prop_types_1.default.number]),
    type: prop_types_1.default.oneOf(['Feature']).isRequired,
    geometry: prop_types_1.default.oneOfType([GeometryType, GeometryCollectionType]).isRequired,
    properties: prop_types_1.default.object
});
exports.FeatureType = FeatureType;
var FeatureCollectionType = prop_types_1.default.shape({
    type: prop_types_1.default.oneOf(['FeatureCollection']).isRequired,
    features: prop_types_1.default.arrayOf(FeatureType).isRequired
});
exports.FeatureCollectionType = FeatureCollectionType;
