// rc-leaflet/lib/components/Point/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var leaflet_1 = __importDefault(require("leaflet"));
var PropTypes_1 = require("../../util/PropTypes");
var Context_1 = __importDefault(require("./Context"));
var InteractiveLayer_1 = __importDefault(require("../InteractiveLayer"));
var Content_1 = __importDefault(require("./Content"));
var DivIcon_1 = require("../DivIcon");
var creator_1 = require("../DivIcon/creator");
var Point = /** @class */ (function (_super) {
    __extends(Point, _super);
    function Point(props, context) {
        var _this = _super.call(this, props, context) || this;
        _this.state = {
            instance: _this.instance,
            icon: props.icon
        };
        return _this;
    }
    Point.getDerivedStateFromProps = function (nextProps, prevState) {
        var nextIcon = nextProps.icon;
        var instance = prevState.instance, icon = prevState.icon;
        if (nextIcon !== icon) {
            if (nextIcon instanceof leaflet_1.default.DivIcon) {
                DivIcon_1.keepPrevHTML(instance, nextIcon);
            }
            return { icon: nextIcon };
        }
        return null;
    };
    Point.prototype.createInstance = function (props) {
        var position = props.position, _a = props.icon, icon = _a === void 0 ? creator_1.defaultIcon : _a, options = __rest(props, ["position", "icon"]);
        return leaflet_1.default.marker(position, __assign({}, options, { icon: icon }));
    };
    Point.prototype.componentDidUpdate = function (prevProps) {
        var prevPosition = prevProps.position, prevZIndexOffset = prevProps.zIndexOffset, prevOpacity = prevProps.opacity;
        var _a = this.props, position = _a.position, zIndexOffset = _a.zIndexOffset, opacity = _a.opacity;
        var point = this.instance;
        if (position !== prevPosition) {
            point.setLatLng(position);
        }
        if (zIndexOffset !== prevZIndexOffset) {
            point.setZIndexOffset(zIndexOffset);
        }
        if (opacity !== prevOpacity) {
            point.setOpacity(opacity);
        }
        _super.prototype.componentDidUpdate.call(this, prevProps);
    };
    Point.prototype.render = function () {
        return (react_1.default.createElement(Context_1.default.Provider, { value: this.state }, _super.prototype.render.call(this)));
    };
    Point.propTypes = __assign({}, InteractiveLayer_1.default.propTypes, { position: PropTypes_1.Point.isRequired, zIndexOffset: prop_types_1.default.number, opacity: prop_types_1.default.number, clickable: prop_types_1.default.bool, draggable: prop_types_1.default.bool, keyboard: prop_types_1.default.bool, title: prop_types_1.default.string, alt: prop_types_1.default.string, riseOnHover: prop_types_1.default.bool, riseOffset: prop_types_1.default.number, autoPan: prop_types_1.default.bool, autoPanSpeed: prop_types_1.default.number, autoPanPadding: PropTypes_1.Pixel });
    Point.Content = Content_1.default;
    return Point;
}(InteractiveLayer_1.default));
exports.default = Point;
